File: algorithms/0210. Course Schedule II.cpp
class Solution 
{
public:
    vector<vector<int>> g;
    vector<int> c;
    vector<int> ans;
    
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) 
    {
        c = vector<int>(numCourses);
        g = vector<vector<int>>(numCourses, vector<int>());
        
        for (vector<int> prerequisite : prerequisites)
        {
            g[prerequisite[1]].push_back(prerequisite[0]);
        }
        
        for (int i = 0; i < numCourses; ++i)
        {
            if (!c[i])
            {
                if (!dfs(i))
                {
                    return {};
                }
            }
        }
        
        reverse(ans.begin(), ans.end());
        
        return ans;
    }
    
    bool dfs(int v)
    {
        c[v] = 1;
        
        for (int u : g[v])
        {
            if (!c[u])
            {
                if (!dfs(u))
                {
                    return false;
                }
            }
            else if (c[u] == 1)
            {
                return false;
            }
        }
        
        c[v] = 2;
        
        ans.push_back(v);
        
        return true;
    }
};

File: 0210. Course Schedule II.cs
public class Solution
{
    public int[] FindOrder(int numCourses, int[][] prerequisites)
    {
        var order = new List<int>();
        var states = new int[numCourses]; // 0 - not visited, 1 - in, 2 - out
        var graph = Enumerable.Range(0, numCourses).Select(x => new List<int>()).ToList();

        foreach (var prerequisite in prerequisites)
        {
            graph[prerequisite[0]].Add(prerequisite[1]);
        }

        for (int course = 0; course < numCourses; ++course)
        {
            if (states[course] == 0)
            {
                if (!TrySort(course, graph, states, order))
                {
                    return [];
                }
            }
        }

        return order.ToArray();
    }

    private bool TrySort(
        int v,
        List<List<int>> graph,
        int[] states,
        List<int> order)
    {
        if (states[v] == 0)
        {
            states[v] = 1;

            foreach (var u in graph[v])
            {
                if (!TrySort(u, graph, states, order))
                {
                    return false;
                }
            }

            states[v] = 2;
            order.Add(v);
        }
        else if (states[v] == 1)
        {
            return false;
        }
        
        return true;
    }
}